File: python/02_analysis.py
"""Player value from lineup data: possession-weighted ridge regression.

The model
---------
Every five-man lineup's net scoring rate (points scored minus allowed, per 100
possessions) is modeled as the sum of five player effects plus an intercept:

    net_100_i  ~  intercept + sum_{p in lineup i} beta_p        (weight: POSS_i)

Ordinary least squares on this design is hopeless: lineups are massively
collinear (stars share the floor with the same teammates all season), and
low-possession lineups have wild observed rates. Ridge regularization shrinks
every player toward zero (league average) in proportion to how little evidence
there is - the standard solution in the public literature (see README,
"Where this sits in the literature"). The penalty is chosen by 5-fold
cross-validation with deterministic folds so the R implementation can
reproduce the choice exactly.

Players below a possession threshold are pooled into a single "replacement"
column: their individual effects are unidentifiable, and pooling them is more
honest than reporting noise shrunk toward zero.

What this is NOT: stint-level RAPM with opponent adjustment. Season lineup
aggregates do not record who the opponents were, so strength of opposition is
not controlled. The README's limitations section says so plainly.

Inputs : data/raw/*.parquet          (from 01_harvest_lineups.py)
Outputs: output/player_values.csv    per-player estimate + bootstrap CI
         output/lineup_table.csv     high-possession lineup ratings
         output/cv_curve.csv         cross-validation curve
         output/validation.csv      external checks (see 04_findings.py)
         figures/*.png + *.html

Run: python python/02_analysis.py
"""
from __future__ import annotations

from pathlib import Path

import numpy as np
import plotly.graph_objects as go
import polars as pl

ROOT = Path(__file__).resolve().parents[1]
RAW_ROOT = ROOT / "data" / "raw"

BLUE, ORANGE, AQUA = "#2a78d6", "#eb6834", "#1baf7a"
INK, INK2, MUTED = "#0b0b0b", "#52514e", "#898781"
GRID, AXIS, SURFACE = "#e1e0d9", "#c3c2b7", "#fcfcfb"

MIN_POSS = 300          # below this a player is pooled into "replacement"
LAMBDAS = [100.0, 200.0, 400.0, 800.0, 1600.0, 3200.0, 6400.0]
N_FOLDS = 5
BOOT_REPS = 500
SEED = 2026


def base_layout(title: str, subtitle: str, x_title: str, y_title: str) -> go.Layout:
    return go.Layout(
        title=dict(text=f"<b>{title}</b><br><span style='font-size:12px;color:{INK2}'>"
                        f"{subtitle}</span>",
                   font=dict(size=17, color=INK), x=0, xanchor="left"),
        paper_bgcolor=SURFACE, plot_bgcolor=SURFACE,
        font=dict(family="system-ui, -apple-system, Segoe UI, sans-serif",
                  size=12, color=INK2),
        xaxis=dict(title=dict(text=x_title, font=dict(color=INK2, size=12)),
                   showgrid=True, gridcolor=GRID, gridwidth=0.5, zeroline=False,
                   linecolor=AXIS, tickfont=dict(color=MUTED, size=11)),
        yaxis=dict(title=dict(text=y_title, font=dict(color=INK2, size=12)),
                   showgrid=True, gridcolor=GRID, gridwidth=0.5, zeroline=False,
                   linecolor=AXIS, tickfont=dict(color=MUTED, size=11)),
        showlegend=False, margin=dict(l=80, r=120, t=85, b=55),
        width=900, height=520,
    )


def save(fig: go.Figure, stem: str) -> None:
    fig.write_image(FIG / f"{stem}.png", scale=2)
    fig.write_html(FIG / f"{stem}.html", include_plotlyjs="cdn")


# ------------------------------------------------------------------ loading --
def load_lineups(raw: Path) -> tuple[pl.DataFrame, pl.DataFrame]:
    """Returns (model_frame, full_frame).

    The model frame drops zero-possession lineups (net_100 is undefined for
    them - they are defensive-only micro-stints, e.g. on the floor solely for
    opponent free throws). The FULL frame keeps them, because the external
    coverage validation must reconstruct team totals exactly, and the dropped
    stints carry real plus-minus points.
    """
    base = pl.concat([pl.read_parquet(f) for f in sorted(raw.glob("lineups_base_*.parquet"))])
    adv = pl.concat([pl.read_parquet(f) for f in sorted(raw.glob("lineups_advanced_*.parquet"))])

    df = (base.select("GROUP_ID", "TEAM_ID", "TEAM_ABBREVIATION", "GROUP_NAME",
                      "MIN", "PLUS_MINUS")
          .join(adv.select("GROUP_ID", "TEAM_ID", "POSS", "NET_RATING"),
                on=["GROUP_ID", "TEAM_ID"], how="inner"))

    full = df
    zero = df.filter(pl.col("POSS") <= 0)
    if zero.height:
        print(f"  model excludes {zero.height} zero-possession lineups "
              f"(defensive-only micro-stints, {zero['PLUS_MINUS'].sum():+.0f} pts "
              f"kept in the coverage validation)")
    df = (df.filter(pl.col("POSS") > 0)
          .with_columns(
              net_100=100.0 * pl.col("PLUS_MINUS") / pl.col("POSS"),
              player_ids=pl.col("GROUP_ID").str.strip_chars("-").str.split("-"),
          )
          # Deterministic row order: the CV fold assignment and the bootstrap
          # both index rows, and R must be able to reproduce them.
          .sort(["GROUP_ID", "TEAM_ID"]))
    assert (df["player_ids"].list.len() == 5).all()
    return df, full


def player_names(raw: Path) -> pl.DataFrame:
    ps = pl.read_parquet(raw / "player_stats.parquet")
    return ps.select(
        player_id=pl.col("PLAYER_ID").cast(pl.Utf8),
        player=pl.col("PLAYER_NAME"),
        team=pl.col("TEAM_ABBREVIATION"),
        season_min=pl.col("MIN"),
        season_pm=pl.col("PLUS_MINUS"),
    )


# ------------------------------------------------------------- design matrix --
def build_design(df: pl.DataFrame) -> tuple[np.ndarray, np.ndarray, np.ndarray, list[str]]:
    """Rows: lineups. Columns: intercept, kept players (sorted id), replacement."""
    poss_by_player: dict[str, float] = {}
    for ids, poss in zip(df["player_ids"].to_list(), df["POSS"].to_list()):
        for pid in ids:
            poss_by_player[pid] = poss_by_player.get(pid, 0.0) + poss

    kept = sorted(pid for pid, p in poss_by_player.items() if p >= MIN_POSS)
    col_of = {pid: j for j, pid in enumerate(kept)}

    n, p = df.height, len(kept)
    X = np.zeros((n, p + 2))          # [intercept | players | replacement]
    X[:, 0] = 1.0
    for i, ids in enumerate(df["player_ids"].to_list()):
        for pid in ids:
            j = col_of.get(pid)
            if j is None:
                X[i, p + 1] += 1.0    # replacement count
            else:
                X[i, 1 + j] = 1.0

    y = df["net_100"].to_numpy().astype(float)
    w = df["POSS"].to_numpy().astype(float)
    return X, y, w, kept


def ridge_fit(X: np.ndarray, y: np.ndarray, w: np.ndarray, lam: float) -> np.ndarray:
    """Closed-form weighted ridge; intercept (column 0) unpenalized."""
    Xw = X * w[:, None]
    A = X.T @ Xw
    pen = np.full(X.shape[1], lam)
    pen[0] = 0.0
    A[np.diag_indices_from(A)] += pen
    return np.linalg.solve(A, Xw.T @ y)


def cv_lambda(X: np.ndarray, y: np.ndarray, w: np.ndarray) -> tuple[float, pl.DataFrame]:
    """Deterministic 5-fold CV (fold = row index mod N_FOLDS on sorted rows)."""
    folds = np.arange(len(y)) % N_FOLDS
    rows = []
    for lam in LAMBDAS:
        sse = wsum = 0.0
        for k in range(N_FOLDS):
            tr, te = folds != k, folds == k
            beta = ridge_fit(X[tr], y[tr], w[tr], lam)
            resid = y[te] - X[te] @ beta
            sse += float(np.sum(w[te] * resid ** 2))
            wsum += float(np.sum(w[te]))
        rows.append({"lam": lam, "cv_wmse": sse / wsum})
    curve = pl.DataFrame(rows)
    best = float(curve.sort("cv_wmse")["lam"][0])
    return best, curve


# ------------------------------------------------------------------ main -----
def run(season: str) -> int:
    raw = RAW_ROOT / season
    out = ROOT / "output" / season
    fig_dir = ROOT / "figures" / season
    out.mkdir(parents=True, exist_ok=True)
    fig_dir.mkdir(parents=True, exist_ok=True)
    global OUT, FIG
    OUT, FIG = out, fig_dir

    print(f"=== {season} ===")
    df, df_full = load_lineups(raw)
    names = player_names(raw)
    print(f"{df.height:,} lineups, {df['TEAM_ID'].n_unique()} teams, "
          f"{df['POSS'].sum():,.0f} possessions")

    X, y, w, kept = build_design(df)
    print(f"design: {X.shape[0]:,} x {X.shape[1]} "
          f"({len(kept)} players >= {MIN_POSS} poss, + intercept + replacement)")

    best_lam, curve = cv_lambda(X, y, w)
    curve.write_csv(OUT / "cv_curve.csv")
    print(f"lambda by 5-fold CV: {best_lam:g}")

    beta = ridge_fit(X, y, w, best_lam)

    # Bootstrap over lineups at the chosen lambda. Resampling rows treats the
    # lineup-season as the sampling unit; the CI reflects how much each player's
    # estimate depends on which lineups happened to exist.
    rng = np.random.default_rng(SEED)
    boots = np.empty((BOOT_REPS, X.shape[1]))
    n = X.shape[0]
    for b in range(BOOT_REPS):
        idx = rng.integers(0, n, n)
        boots[b] = ridge_fit(X[idx], y[idx], w[idx], best_lam)
    lo, hi = np.quantile(boots, [0.025, 0.975], axis=0)

    poss_kept = {pid: 0.0 for pid in kept}
    for ids, poss in zip(df["player_ids"].to_list(), df["POSS"].to_list()):
        for pid in ids:
            if pid in poss_kept:
                poss_kept[pid] += poss

    values = (pl.DataFrame({
        "player_id": kept,
        "rapm_100": beta[1:1 + len(kept)],
        "ci_lo": lo[1:1 + len(kept)],
        "ci_hi": hi[1:1 + len(kept)],
        "poss": [poss_kept[pid] for pid in kept],
    })
        .join(names, on="player_id", how="left")
        .sort("rapm_100", descending=True)
        .with_row_index("rank", offset=1))
    values.write_csv(OUT / "player_values.csv")

    intercept, repl = beta[0], beta[-1]
    print(f"intercept {intercept:+.2f} (league avg), replacement effect {repl:+.2f}/100")

    # High-possession lineup table: the lineup-level deliverable.
    lineups = (df.filter(pl.col("POSS") >= 1000)
               .select("TEAM_ABBREVIATION", "GROUP_NAME", "MIN", "POSS",
                       "PLUS_MINUS", "net_100", "NET_RATING")
               .sort("net_100", descending=True))
    lineups.write_csv(OUT / "lineup_table.csv")

    # ---- external validation ------------------------------------------------
    teams = pl.read_parquet(raw / "team_stats.parquet")
    cov = (df_full.group_by("TEAM_ID")
           .agg(lineup_min=pl.col("MIN").sum(), lineup_pm=pl.col("PLUS_MINUS").sum())
           .join(teams.select("TEAM_ID", team_min=pl.col("MIN"),
                              team_pm=pl.col("PLUS_MINUS")), on="TEAM_ID")
           .with_columns(
               min_ratio=pl.col("lineup_min") / pl.col("team_min"),
               pm_err=(pl.col("lineup_pm") - pl.col("team_pm")).abs(),
           ))
    min_ratio_worst = float((cov["min_ratio"] - 1.0).abs().max())
    pm_err_max = float(cov["pm_err"].max())

    # our net_100 vs the NBA's own NET_RATING for the same lineups (they use
    # separate off/def possession denominators; agreement should be near 1)
    big = df.filter(pl.col("POSS") >= 200)
    r_net = float(np.corrcoef(big["net_100"].to_numpy(),
                              big["NET_RATING"].to_numpy())[0, 1])

    # face validity: rank-correlation with raw season plus-minus. Should be
    # clearly positive but NOT 1 - the daylight between them is the teammate
    # adjustment doing its work.
    vv = values.drop_nulls("season_pm").with_columns(
        pm_per_min=pl.col("season_pm") / pl.col("season_min"))
    def spearman(a: np.ndarray, b: np.ndarray) -> float:
        ra = np.argsort(np.argsort(a)).astype(float)
        rb = np.argsort(np.argsort(b)).astype(float)
        return float(np.corrcoef(ra, rb)[0, 1])
    rho_pm = spearman(vv["rapm_100"].to_numpy(), vv["pm_per_min"].to_numpy())

    validation = pl.DataFrame([
        {"check": "lineup minutes reconstruct team minutes",
         "value": min_ratio_worst, "threshold": 0.005, "pass": min_ratio_worst <= 0.005},
        {"check": "lineup plus-minus reconstructs team plus-minus (max abs pts)",
         "value": pm_err_max, "threshold": 1.0, "pass": pm_err_max <= 1.0},
        {"check": "net_100 vs NBA NET_RATING correlation (POSS>=200)",
         "value": r_net, "threshold": 0.97, "pass": r_net >= 0.97},
        {"check": "Spearman(RAPM, raw plus-minus per minute)",
         "value": rho_pm, "threshold": 0.5, "pass": rho_pm >= 0.5},
    ])
    validation.write_csv(OUT / "validation.csv")
    print(validation)
    if not validation["pass"].all():
        print("VALIDATION FAILED")
        return 1

    # ---- figures ------------------------------------------------------------
    top = values.head(15)
    bot = values.tail(15).sort("rapm_100")
    show = pl.concat([top, bot]).sort("rapm_100")
    fig = go.Figure(layout=base_layout(
        f"Regularized player value from {season} lineup data",
        f"Points per 100 possessions vs league average · ridge, lambda={best_lam:g} by CV · "
        f"95% bootstrap CI · players with {MIN_POSS}+ possessions",
        "Net points per 100 possessions (vs average)", ""))
    fig.update_layout(height=760, margin=dict(l=170, r=90, t=95, b=55))
    fig.add_vline(x=0, line=dict(color=AXIS, width=1))
    fig.add_trace(go.Scatter(
        x=show["rapm_100"].to_list(), y=show["player"].to_list(),
        mode="markers",
        error_x=dict(type="data", symmetric=False,
                     array=(show["ci_hi"] - show["rapm_100"]).to_list(),
                     arrayminus=(show["rapm_100"] - show["ci_lo"]).to_list(),
                     color=MUTED, thickness=1),
        marker=dict(size=8, color=[BLUE if v > 0 else ORANGE for v in show["rapm_100"]]),
        hovertemplate="%{y}: %{x:+.2f} per 100<extra></extra>"))
    save(fig, "fig1_player_values")  # season dir keeps them apart

    fig = go.Figure(layout=base_layout(
        "Ridge shrinks low-evidence players toward average",
        "Each dot is a player: naive per-possession impact vs the regularized estimate",
        "Naive on-court net per 100 (possession-weighted mean of lineups)",
        "Ridge estimate per 100"))
    naive = []
    for pid in kept:
        mask = np.array([pid in ids for ids in df["player_ids"].to_list()])
        naive.append(float(np.average(y[mask], weights=w[mask])))
    poss_arr = np.array([poss_kept[pid] for pid in kept])
    fig.add_trace(go.Scatter(
        x=naive, y=beta[1:1 + len(kept)], mode="markers",
        marker=dict(size=np.clip(np.sqrt(poss_arr) / 6, 3, 14), color=BLUE,
                    opacity=0.55, line=dict(width=0)),
        hoverinfo="skip"))
    lim = max(map(abs, naive)) * 1.05
    fig.add_trace(go.Scatter(x=[-lim, lim], y=[-lim, lim], mode="lines",
                             line=dict(color=AXIS, width=1, dash="dot"),
                             hoverinfo="skip"))
    fig.add_annotation(x=lim * 0.95, y=lim * 0.8, text="no shrinkage line",
                       showarrow=False, font=dict(color=MUTED, size=11), xanchor="right")
    save(fig, "fig2_shrinkage")

    fig = go.Figure(layout=base_layout(
        "Penalty chosen by cross-validation",
        "Possession-weighted MSE across 5 deterministic folds",
        "Ridge penalty (log scale)", "CV weighted MSE"))
    fig.add_trace(go.Scatter(x=curve["lam"].to_list(), y=curve["cv_wmse"].to_list(),
                             mode="lines+markers", line=dict(color=BLUE, width=2),
                             marker=dict(size=8, color=BLUE),
                             hovertemplate="lambda %{x:g}: %{y:.2f}<extra></extra>"))
    fig.update_xaxes(type="log")
    fig.add_vline(x=best_lam, line=dict(color=ORANGE, width=1.5, dash="dot"))
    save(fig, "fig3_cv_curve")

    print(f"\nTop 10:")
    for r in values.head(10).iter_rows(named=True):
        print(f"  {r['rank']:>3} {r['player']:<28} {r['rapm_100']:+.2f} "
              f"[{r['ci_lo']:+.2f}, {r['ci_hi']:+.2f}]  ({r['poss']:,.0f} poss)")
    print(f"\nWrote {OUT} and {FIG}")
    return 0


def main() -> int:
    import argparse
    ap = argparse.ArgumentParser()
    ap.add_argument("--seasons", nargs="+",
                    default=sorted(p.name for p in RAW_ROOT.iterdir() if p.is_dir()))
    args = ap.parse_args()
    rc = 0
    for season in args.seasons:
        rc |= run(season)
    return rc


if __name__ == "__main__":
    raise SystemExit(main())

File: R/02_analysis.R
# Player value from lineup data --- R / tidyverse implementation
#
# Independent implementation of python/02_analysis.py: same model (possession-
# weighted ridge with an unpenalized intercept and a pooled replacement column),
# same deterministic 5-fold CV, written from the definitions so that
# python/03_reconcile.py is a real check.
#
# Reconciliation notes, the two places where R and Python silently diverge:
#   * Row and player ordering use C-locale byte order (method = "radix"),
#     matching polars' string sort. R's default locale-aware sort does NOT.
#   * The CV fold of row i (0-based) is i %% 5 on the sorted rows.
#
# Outputs per season: output/<season>/player_values_r.csv, cv_curve_r.csv,
#                     lineup_table_r.csv, validation_r.csv, bootstrap CIs
#                     figures/<season>/fig1_player_values_r.png
#
# Run:  Rscript R/02_analysis.R

suppressPackageStartupMessages({
  library(nanoparquet)
  library(dplyr)
  library(tidyr)
  library(stringr)
  library(purrr)
  library(readr)
  library(ggplot2)
})

ROOT <- normalizePath(file.path(dirname(sub("^--file=", "", grep("^--file=", commandArgs(), value = TRUE)[1])), ".."))
if (is.na(ROOT) || !dir.exists(ROOT)) ROOT <- normalizePath(".")

MIN_POSS <- 300
LAMBDAS <- c(100, 200, 400, 800, 1600, 3200, 6400)
N_FOLDS <- 5
BOOT_REPS <- 500

PAL <- c(blue = "#2a78d6", orange = "#eb6834")
INK <- "#0b0b0b"; INK2 <- "#52514e"; MUTED <- "#898781"
GRID <- "#e1e0d9"; AXIS <- "#c3c2b7"; SURFACE <- "#fcfcfb"

# ridge_fit() and cv_lambda() live in R/functions.R and are unit-tested.
source(file.path(ROOT, "R", "functions.R"))

run_season <- function(season) {
  raw <- file.path(ROOT, "data", "raw", season)
  out_dir <- file.path(ROOT, "output", season)
  fig_dir <- file.path(ROOT, "figures", season)
  dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
  dir.create(fig_dir, showWarnings = FALSE, recursive = TRUE)

  message("=== ", season, " ===")

  read_all <- function(pattern) {
    files <- sort(list.files(raw, pattern = pattern, full.names = TRUE))
    map_dfr(files, read_parquet)
  }
  base <- read_all("^lineups_base_.*\\.parquet$")
  adv <- read_all("^lineups_advanced_.*\\.parquet$")

  df_full <- base %>%
    select(GROUP_ID, TEAM_ID, TEAM_ABBREVIATION, GROUP_NAME, MIN, PLUS_MINUS) %>%
    inner_join(adv %>% select(GROUP_ID, TEAM_ID, POSS, NET_RATING),
               by = c("GROUP_ID", "TEAM_ID"))
  # The model drops zero-possession lineups (net_100 undefined: defensive-only
  # micro-stints), but the coverage validation keeps them - their plus-minus
  # points are real and team totals must reconstruct exactly.
  df <- df_full %>%
    filter(POSS > 0) %>%
    mutate(net_100 = 100 * PLUS_MINUS / POSS)

  # C-locale byte-order sort to match polars exactly.
  df <- df[order(df$GROUP_ID, df$TEAM_ID, method = "radix"), ]

  id_lists <- str_split(str_remove_all(df$GROUP_ID, "^-|-$"), "-")
  stopifnot(all(lengths(id_lists) == 5))

  message(sprintf("%s lineups, %d teams, %s possessions",
                  format(nrow(df), big.mark = ","), n_distinct(df$TEAM_ID),
                  format(round(sum(df$POSS)), big.mark = ",")))

  # ---- design matrix -------------------------------------------------------
  poss_by_player <- tapply(rep(df$POSS, each = 5), unlist(id_lists), sum)
  kept <- sort(names(poss_by_player)[poss_by_player >= MIN_POSS], method = "radix")
  col_of <- setNames(seq_along(kept), kept)

  n <- nrow(df); p <- length(kept)
  X <- matrix(0, n, p + 2)   # [intercept | players | replacement]
  X[, 1] <- 1
  for (i in seq_len(n)) {
    for (pid in id_lists[[i]]) {
      j <- col_of[pid]
      if (is.na(j)) X[i, p + 2] <- X[i, p + 2] + 1 else X[i, 1 + j] <- 1
    }
  }
  y <- df$net_100
  w <- df$POSS

  # ---- CV over the lambda grid (deterministic folds) -----------------------
  cv <- cv_lambda(X, y, w, LAMBDAS, N_FOLDS)
  curve <- cv$curve
  write_csv(curve, file.path(out_dir, "cv_curve_r.csv"))
  best_lam <- cv$best
  message("lambda by 5-fold CV: ", best_lam)

  beta <- ridge_fit(X, y, w, best_lam)

  # ---- bootstrap over lineups (R RNG; reconciled by CI overlap) ------------
  set.seed(2026)
  boots <- matrix(NA_real_, BOOT_REPS, ncol(X))
  for (b in seq_len(BOOT_REPS)) {
    idx <- sample.int(n, n, replace = TRUE)
    boots[b, ] <- ridge_fit(X[idx, , drop = FALSE], y[idx], w[idx], best_lam)
  }
  ci <- apply(boots, 2, quantile, probs = c(0.025, 0.975))

  poss_kept <- poss_by_player[kept]

  names_tbl <- read_parquet(file.path(raw, "player_stats.parquet")) %>%
    transmute(player_id = as.character(PLAYER_ID), player = PLAYER_NAME,
              team = TEAM_ABBREVIATION, season_min = MIN, season_pm = PLUS_MINUS)

  values <- tibble(
    player_id = kept,
    rapm_100 = beta[2:(1 + p)],
    ci_lo = ci[1, 2:(1 + p)],
    ci_hi = ci[2, 2:(1 + p)],
    poss = as.numeric(poss_kept),
  ) %>%
    left_join(names_tbl, by = "player_id") %>%
    arrange(desc(rapm_100)) %>%
    mutate(rank = row_number(), .before = 1)
  write_csv(values, file.path(out_dir, "player_values_r.csv"))

  df %>%
    filter(POSS >= 1000) %>%
    select(TEAM_ABBREVIATION, GROUP_NAME, MIN, POSS, PLUS_MINUS, net_100, NET_RATING) %>%
    arrange(desc(net_100)) %>%
    write_csv(file.path(out_dir, "lineup_table_r.csv"))

  # ---- external validation (same checks as Python) -------------------------
  teams <- read_parquet(file.path(raw, "team_stats.parquet"))
  cov <- df_full %>%
    group_by(TEAM_ID) %>%
    summarise(lineup_min = sum(MIN), lineup_pm = sum(PLUS_MINUS), .groups = "drop") %>%
    inner_join(teams %>% select(TEAM_ID, team_min = MIN, team_pm = PLUS_MINUS),
               by = "TEAM_ID")
  min_ratio_worst <- max(abs(cov$lineup_min / cov$team_min - 1))
  pm_err_max <- max(abs(cov$lineup_pm - cov$team_pm))

  big <- df %>% filter(POSS >= 200)
  r_net <- cor(big$net_100, big$NET_RATING)

  vv <- values %>% filter(!is.na(season_pm)) %>%
    mutate(pm_per_min = season_pm / season_min)
  rho_pm <- cor(vv$rapm_100, vv$pm_per_min, method = "spearman")

  validation <- tibble(
    check = c("lineup minutes reconstruct team minutes",
              "lineup plus-minus reconstructs team plus-minus (max abs pts)",
              "net_100 vs NBA NET_RATING correlation (POSS>=200)",
              "Spearman(RAPM, raw plus-minus per minute)"),
    value = c(min_ratio_worst, pm_err_max, r_net, rho_pm),
    threshold = c(0.005, 1.0, 0.97, 0.5),
    pass = c(min_ratio_worst <= 0.005, pm_err_max <= 1.0, r_net >= 0.97, rho_pm >= 0.5)
  )
  write_csv(validation, file.path(out_dir, "validation_r.csv"))
  print(as.data.frame(validation))
  stopifnot(all(validation$pass))

  # ---- figure --------------------------------------------------------------
  show <- bind_rows(head(values, 15), tail(values, 15)) %>%
    mutate(player = factor(player, levels = player[order(rapm_100)]))
  p1 <- ggplot(show, aes(rapm_100, player)) +
    geom_vline(xintercept = 0, colour = AXIS, linewidth = 0.3) +
    geom_errorbarh(aes(xmin = ci_lo, xmax = ci_hi), height = 0,
                   colour = MUTED, linewidth = 0.4) +
    geom_point(aes(colour = rapm_100 > 0), size = 2.2, show.legend = FALSE) +
    scale_colour_manual(values = c(`TRUE` = PAL[["blue"]], `FALSE` = PAL[["orange"]])) +
    labs(
      title = paste0("Regularized player value from ", season, " lineup data"),
      subtitle = paste0("Points per 100 possessions vs league average · ridge, lambda=",
                        best_lam, " by CV · 95% bootstrap CI"),
      x = "Net points per 100 possessions (vs average)", y = NULL
    ) +
    theme_minimal(base_size = 11) +
    theme(
      plot.background = element_rect(fill = SURFACE, colour = NA),
      panel.background = element_rect(fill = SURFACE, colour = NA),
      panel.grid.major = element_line(colour = GRID, linewidth = 0.3),
      panel.grid.minor = element_blank(),
      axis.text = element_text(colour = MUTED, size = 8.5),
      axis.title = element_text(colour = INK2, size = 9),
      plot.title = element_text(colour = INK, face = "bold", size = 12),
      plot.subtitle = element_text(colour = INK2, size = 9)
    )
  ggsave(file.path(fig_dir, "fig1_player_values_r.png"), p1,
         width = 7.5, height = 6.3, dpi = 200)

  message("Top 5: ", paste(head(values$player, 5), collapse = ", "))
}

seasons <- list.dirs(file.path(ROOT, "data", "raw"), recursive = FALSE, full.names = FALSE)
for (s in sort(seasons)) run_season(s)
message("R analysis complete.")
